// @shockpkg/swf-projector/cjs/util/internal/windows/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IMAGE_SCN_MEM_WRITE = exports.IMAGE_SCN_MEM_SHARED = exports.IMAGE_SCN_MEM_READ = exports.IMAGE_SCN_MEM_EXECUTE = exports.IMAGE_SCN_CNT_UNINITIALIZED_DATA = exports.IMAGE_SCN_CNT_INITIALIZED_DATA = exports.IMAGE_SCN_CNT_CODE = exports.IDD_RESOURCE = exports.IDD_RESERVED = exports.IDD_EXCEPTION = exports.IDD_BASE_RELOCATION = void 0;
// IMAGE_DATA_DIRECTORY indexes.
const IDD_RESOURCE = exports.IDD_RESOURCE = 2;
const IDD_EXCEPTION = exports.IDD_EXCEPTION = 3;
const IDD_BASE_RELOCATION = exports.IDD_BASE_RELOCATION = 5;
const IDD_RESERVED = exports.IDD_RESERVED = 15;

// IMAGE_SECTION_HEADER characteristics.
const IMAGE_SCN_CNT_CODE = exports.IMAGE_SCN_CNT_CODE = 0x00000020;
const IMAGE_SCN_CNT_INITIALIZED_DATA = exports.IMAGE_SCN_CNT_INITIALIZED_DATA = 0x00000040;
const IMAGE_SCN_CNT_UNINITIALIZED_DATA = exports.IMAGE_SCN_CNT_UNINITIALIZED_DATA = 0x00000080;
const IMAGE_SCN_MEM_SHARED = exports.IMAGE_SCN_MEM_SHARED = 0x10000000;
const IMAGE_SCN_MEM_EXECUTE = exports.IMAGE_SCN_MEM_EXECUTE = 0x20000000;
const IMAGE_SCN_MEM_READ = exports.IMAGE_SCN_MEM_READ = 0x40000000;
const IMAGE_SCN_MEM_WRITE = exports.IMAGE_SCN_MEM_WRITE = 0x80000000;
