// @shockpkg/swf-projector/cjs/util/internal/mac/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VM_PROT_READ = exports.SEG_TEXT = exports.SEG_LINKEDIT = exports.SECT_TEXT = exports.MH_MAGIC_64 = exports.MH_MAGIC = exports.MH_CIGAM_64 = exports.MH_CIGAM = exports.LC_SYMTAB = exports.LC_SEGMENT_SPLIT_INFO = exports.LC_SEGMENT_64 = exports.LC_SEGMENT = exports.LC_REQ_DYLD = exports.LC_LINKER_OPTIMIZATION_HINT = exports.LC_FUNCTION_STARTS = exports.LC_DYSYMTAB = exports.LC_DYLIB_CODE_SIGN_DRS = exports.LC_DYLD_INFO_ONLY = exports.LC_DYLD_INFO = exports.LC_DYLD_EXPORTS_TRIE = exports.LC_DYLD_CHAINED_FIXUPS = exports.LC_DATA_IN_CODE = exports.LC_CODE_SIGNATURE = exports.FAT_MAGIC = exports.CPU_TYPE_X86_64 = exports.CPU_TYPE_POWERPC64 = exports.CPU_TYPE_POWERPC = exports.CPU_TYPE_I386 = exports.CPU_TYPE_ARM64 = void 0;
const VM_PROT_READ = exports.VM_PROT_READ = 0x1;
const FAT_MAGIC = exports.FAT_MAGIC = 0xcafebabe;
const MH_MAGIC = exports.MH_MAGIC = 0xfeedface;
const MH_CIGAM = exports.MH_CIGAM = 0xcefaedfe;
const MH_MAGIC_64 = exports.MH_MAGIC_64 = 0xfeedfacf;
const MH_CIGAM_64 = exports.MH_CIGAM_64 = 0xcffaedfe;
const LC_REQ_DYLD = exports.LC_REQ_DYLD = 0x80000000;
const LC_SEGMENT = exports.LC_SEGMENT = 0x1;
const LC_SYMTAB = exports.LC_SYMTAB = 0x2;
const LC_DYSYMTAB = exports.LC_DYSYMTAB = 0xb;
const LC_SEGMENT_64 = exports.LC_SEGMENT_64 = 0x19;
const LC_CODE_SIGNATURE = exports.LC_CODE_SIGNATURE = 0x1d;
const LC_SEGMENT_SPLIT_INFO = exports.LC_SEGMENT_SPLIT_INFO = 0x1e;
const LC_DYLD_INFO = exports.LC_DYLD_INFO = 0x22;
// eslint-disable-next-line no-bitwise
const LC_DYLD_INFO_ONLY = exports.LC_DYLD_INFO_ONLY = (0x22 | LC_REQ_DYLD) >>> 0;
const LC_FUNCTION_STARTS = exports.LC_FUNCTION_STARTS = 0x26;
const LC_DATA_IN_CODE = exports.LC_DATA_IN_CODE = 0x29;
const LC_DYLIB_CODE_SIGN_DRS = exports.LC_DYLIB_CODE_SIGN_DRS = 0x2b;
const LC_LINKER_OPTIMIZATION_HINT = exports.LC_LINKER_OPTIMIZATION_HINT = 0x2e;
// eslint-disable-next-line no-bitwise
const LC_DYLD_EXPORTS_TRIE = exports.LC_DYLD_EXPORTS_TRIE = (0x33 | LC_REQ_DYLD) >>> 0;
// eslint-disable-next-line no-bitwise
const LC_DYLD_CHAINED_FIXUPS = exports.LC_DYLD_CHAINED_FIXUPS = (0x34 | LC_REQ_DYLD) >>> 0;
const SEG_TEXT = exports.SEG_TEXT = '__TEXT';
const SECT_TEXT = exports.SECT_TEXT = '__text';
const SEG_LINKEDIT = exports.SEG_LINKEDIT = '__LINKEDIT';
const CPU_TYPE_POWERPC = exports.CPU_TYPE_POWERPC = 0x00000012;
const CPU_TYPE_POWERPC64 = exports.CPU_TYPE_POWERPC64 = 0x01000012;
const CPU_TYPE_I386 = exports.CPU_TYPE_I386 = 0x00000007;
const CPU_TYPE_X86_64 = exports.CPU_TYPE_X86_64 = 0x01000007;
const CPU_TYPE_ARM64 = exports.CPU_TYPE_ARM64 = 0x0100000c;
