// @shockpkg/swf-projector/cjs/util/internal/linux/menu64.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.menu64 = exports.PatchMenu64 = void 0;
var _asm = require("./asm.js");
var _menu = require("./menu.js");
/* eslint-disable max-classes-per-file */

/**
 * PatchMenu64 object.
 */
class PatchMenu64 extends _menu.PatchMenu {}

/**
 * Patch objects.
 */
exports.PatchMenu64 = PatchMenu64;
const menu64 = exports.menu64 = [
/**
 * 24.0.0.186 x86_64.
 */
class extends PatchMenu64 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 1,
    find: _asm.MENU_X8664['24-widget-a'],
    replace: _asm.MENU_X8664['24-widget-b']
  }, {
    count: 1,
    find: _asm.MENU_X8664['24-menu-a'],
    replace: _asm.MENU_X8664['24-menu-b']
  }];
},
/**
 * 32.0.0.293 x86_64.
 */
class extends PatchMenu64 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 1,
    find: _asm.MENU_X8664['32-widget-a'],
    replace: _asm.MENU_X8664['32-widget-b']
  }, {
    count: 1,
    find: _asm.MENU_X8664['32-menu-a'],
    replace: _asm.MENU_X8664['32-menu-b']
  }];
}];
