// @shockpkg/swf-projector/cjs/util/internal/linux/menu32.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.menu32 = exports.PatchMenu32 = void 0;
var _asm = require("./asm.js");
var _menu = require("./menu.js");
/* eslint-disable max-classes-per-file */

/**
 * PatchMenu32 object.
 */
class PatchMenu32 extends _menu.PatchMenu {}

/**
 * Patch objects.
 */
exports.PatchMenu32 = PatchMenu32;
const menu32 = exports.menu32 = [
/**
 * 6.0.79.0 i386.
 */
class extends PatchMenu32 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 1,
    find: _asm.MENU_I386['6-widget-a'],
    replace: _asm.MENU_I386['6-widget-b']
  }];
},
/**
 * 9.0.115.0 i386.
 */
class extends PatchMenu32 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 2,
    find: _asm.MENU_I386['9-widget-a'],
    replace: _asm.MENU_I386['9-widget-b']
  }, {
    count: 2,
    find: _asm.MENU_I386['9-menu-a'],
    replace: _asm.MENU_I386['9-menu-b']
  }];
},
/**
 * 10.0.12.36 i386.
 */
class extends PatchMenu32 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 2,
    find: _asm.MENU_I386['10.0-widget-a'],
    replace: _asm.MENU_I386['10.0-widget-b']
  }, {
    count: 1,
    find: _asm.MENU_I386['10.0-menu-1-a'],
    replace: _asm.MENU_I386['10.0-menu-1-b']
  }, {
    count: 1,
    find: _asm.MENU_I386['10.0-menu-2-a'],
    replace: _asm.MENU_I386['10.0-menu-2-b']
  }];
},
/**
 * 10.1.53.64 i386.
 */
class extends PatchMenu32 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 1,
    find: _asm.MENU_I386['10.1-widget-a'],
    replace: _asm.MENU_I386['10.1-widget-b']
  }, {
    count: 1,
    find: _asm.MENU_I386['10.1-menu-a'],
    replace: _asm.MENU_I386['10.1-menu-b']
  }];
},
/**
 * 11.0.1.152 i386.
 */
class extends PatchMenu32 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 1,
    find: _asm.MENU_I386['11.0-widget-a'],
    replace: _asm.MENU_I386['11.0-widget-b']
  }, {
    count: 1,
    find: _asm.MENU_I386['11.0-menu-a'],
    replace: _asm.MENU_I386['11.0-menu-b']
  }];
},
/**
 * 11.2.202.228 i386.
 */
class extends PatchMenu32 {
  /**
   * @inheritDoc
   */
  _spec = [{
    count: 1,
    find: _asm.MENU_I386['11.2-widget-a'],
    replace: _asm.MENU_I386['11.2-widget-b']
  }, {
    count: 1,
    find: _asm.MENU_I386['11.2-menu-a'],
    replace: _asm.MENU_I386['11.2-menu-b']
  }];
}];
