// @shockpkg/plist-dom/cjs/value/integer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValueInteger = void 0;
var _options = require("../options.js");
var _util = require("../util.js");
var _value = require("../value.js");
const MAX_NUM = 0x1fffffffffffff;

/**
 * Assert number is integer.
 *
 * @param value Number value.
 */
function assertInteger(value) {
  if (typeof value === 'number' && !Number.isInteger(value)) {
    // eslint-disable-next-line unicorn/prefer-type-error
    throw new Error(`Value not an integer: ${value}`);
  }
}

/**
 * ValueInteger object.
 */
class ValueInteger extends _value.Value {
  /**
   * Value type.
   */
  static TYPE = 'integer';

  /**
   * Tag names.
   */
  static TAG_NAMES = ['integer'];

  /**
   * Value value.
   */
  value = 0;

  /**
   * ValueInteger constructor.
   *
   * @param value The value.
   */
  constructor(value = 0) {
    super();
    assertInteger(value);
    this.value = value;
  }

  /**
   * Get value as BigInt.
   *
   * @returns Value as a BigInt.
   */
  asBigInt() {
    return BigInt(this.value);
  }

  /**
   * Get value as Number, possibly losing some precision.
   *
   * @returns Value as a Number.
   */
  asNumber() {
    return Number(this.value);
  }

  /**
   * @inheritdoc
   */
  fromXmlElement(element) {
    (0, _util.assertXmlTagName)(element, 'integer');
    const text = (0, _util.xmlElementText)(element)?.nodeValue || '';
    if (!/^[+-]?\d+$/.test(text)) {
      throw new Error(`Invalid integer data: ${text}`);
    }
    const num = +text;
    this.value = num > MAX_NUM || num < -MAX_NUM ? BigInt(text) : num;
  }

  /**
   * @inheritdoc
   */
  toXml(options = null, depth = 0) {
    const p = (options?.indentString ?? _options.INDENT_STRING).repeat(depth);
    const v = this.value;
    assertInteger(v);
    return `${p}<integer>${v}</integer>`;
  }
}
exports.ValueInteger = ValueInteger;
